sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -229,3 +208,27 @@ class EquivalentExpressions {
     sb.toString()
   }
 }
+
+/**
+ * Orders `Expression` by parent/child relations. The child expression is smaller
+ * than parent expression. If there is child-parent relationships among the subexpressions,
+ * we want the child expressions come first than parent expressions, so we can replace
+ * child expressions in parent expressions with subexpression evaluation. Note that
+ * this is not for general expression ordering. For example, two irrelevant or semantically-equal
+ * expressions will be considered as equal by this ordering. But for the usage here, the order of
+ * irrelevant expressions does not matter.
+ */
+class ExpressionContainmentOrdering extends Ordering[Expression] {
+  override def compare(x: Expression, y: Expression): Int = {
+    if (x.find(_.semanticEquals(y)).isDefined) {
+      // `y` is child expression of `x`.
+      1
+    } else if (y.find(_.semanticEquals(x)).isDefined) {
+      // `x` is child expression of `y`.
+      -1
+    } else {
+      // Irrelevant expressions
```

Comment:
ditto. We should mention the semantically-equal expression here.

Reply:
added. thanks.